File: zio-8807/src/mpsc/custom/Basic.scala
package mpsc
package custom

import java.util.concurrent.atomic.AtomicReference

final class Basic[A] extends AtomicReference(List.empty[A]) with Queue[A] {

  @transient private var read = List.empty[A]

  def add(head: A): Unit = {
    var tail = null.asInstanceOf[List[A]]
    var prev = get()
    do {
      tail = prev
      prev = compareAndExchange(tail, head :: tail)
    } while (prev ne tail)
  }

  def poll(): A = {
    var read = this.read
    if (read.isEmpty) {
      read = getAndSet(Nil)
      if (read eq Nil) return null.asInstanceOf[A]
      if (read.tail ne Nil) read = read.reverse
    }
    this.read = read.tail
    read.head
  }
}

File: zio-8807/src/mpsc/vyukov/Basic.java
package mpsc.vyukov;

import mpsc.Queue;

import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;

public final class Basic<A> implements Queue<A> {

	private transient Node read;
	@SuppressWarnings("unused")
	private transient volatile Node write;

	public Basic() {
		read = write = new Node(null);
	}

	public void add(A data) {
		Node next = new Node(data);
		Node prev = WRITE.getAndSet(this, next);
		NEXT.lazySet(prev, next);
	}

	@SuppressWarnings("unchecked")
	public A poll() {
		Node next = read.next;

		if (next == null)
			return null;

		read = next;
		Object data = next.data;
		next.data = null;
		return (A) (data);
	}

	static class Node {
		Object data;
		volatile Node next;

		Node(Object data) {
			this.data = data;
		}

		Node(Object data, Node next) {
			this.data = data;
			this.next = next;
		}
	}

	@SuppressWarnings("rawtypes")
	private static final AtomicReferenceFieldUpdater<Basic, Node> WRITE = AtomicReferenceFieldUpdater
			.newUpdater(Basic.class, Node.class, "write");
	private static final AtomicReferenceFieldUpdater<Node, Node> NEXT = AtomicReferenceFieldUpdater
			.newUpdater(Node.class, Node.class, "next");
}
